### ISSUE TITLE
initializing an array of non-nilable classes

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#20512]
I'd like to be able to create an array of non-nilable classes.
However, creating such an array is actually quite hard.

Here is the obvious code:

``` chapel
class C { }

proc main() {
  var A: [1..0] owned C;
}
```

That fails with `error: cannot default-initialize the array A because it has a non-nilable element type 'owned C'`.

But, that is annoying, because the array has 0 elements, so no element needs to be default-initialized.

We have [`domain.unsafeAssign`](https://chapel-lang.org/docs/language/spec/domains.html#ChapelDomain.unsafeAssignManager) for initializing non-default initializable array elements. Could there be a similar feature for initializing an array? Note that I can't use it to initialize an array (because the array has to exist first in order for me to assign it...)

In the meantime, one can use ugly workarounds like this:

``` chapel
class C { }

proc main() {
  var A: [1..0] owned C = for i in 1..0 do new C();
}
```

or, for an array of borrows:

``` chapel
class C { }

proc main() {
  var dummyC: unmanaged C? = nil;
  var A: [1..0] borrowed C = for i in 1..0 do dummyC!;
}
```

**Comment by bradcray** [comment]
Just to make sure I'm understanding:  The array declared over `[1..0]` here will always be empty because its domain wasn't declared as a variable, so can't be re-assigned, right?  That's not to say you should get this error message, but even if you didn't get it, that the array wouldn't be particularly useful?  I feel like I must be missing something...  Maybe it's just that in a real code, you might have hoisted the `{1..0}` into a domain variable that you could change, but would still need to deal with the initialization of the empty array problem?

> We have [domain.unsafeAssign](https://chapel-lang.org/docs/language/spec/domains.html#ChapelDomain.unsafeAssignManager) for initializing non-default initializable array elements.

Just as a minor comment (fairly unrelated to this issue), I think this feature only works for arrays of non-nilable class elements today, not more general non-default initializable element types (due to the challenge of not having an obvious sentinel value in the more general cases, so the need for more book-keeping vs. relying on `nil` as in the class case).

**Comment by mppf** [comment]
> Just to make sure I'm understanding: The array declared over `[1..0]` here will always be empty because its domain wasn't declared as a variable, so can't be re-assigned, right? That's not to say you should get this error message, but even if you didn't get it, that the array wouldn't be particularly useful? I feel like I must be missing something... Maybe it's just that in a real code, you might have hoisted the `{1..0}` into a domain variable that you could change, but would still need to deal with the initialization of the empty array problem?

Yes, in real code, I would use a separate domain variable that I could mutate later.
